lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,88 +3642,1015 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    auto *dfStruct = linearMapInfo->getLinearMapStruct(origEntry);
+    auto dfStructType =
+        dfStruct->getDeclaredInterfaceType()->getCanonicalType();
+    dfParams.push_back({dfStructType, ParameterConvention::Direct_Guaranteed});
+
+    auto diffName = original->getASTContext()
+                        .getIdentifier("AD__" + original->getName().str() +
+                                       "__differential_" + indices.mangle())
+                        .str();
+    auto diffGenericSig = getAssociatedFunctionGenericSignature(attr, original);
+    auto *diffGenericEnv =
+        diffGenericSig ? diffGenericSig->createGenericEnvironment() : nullptr;
+    auto diffType = SILFunctionType::get(
+        diffGenericSig, origTy->getExtInfo(), origTy->getCoroutineKind(),
+        origTy->getCalleeConvention(), dfParams, {}, dfResults, None,
+        original->getASTContext());
+
+    SILOptFunctionBuilder fb(context.getTransform());
+    // The generated tangent linkage is set to Hidden because generated tangent
+    // are never called cross-module.
+    auto linkage = SILLinkage::Hidden;
+    auto *differential = fb.createFunction(
+        linkage, diffName, diffType, diffGenericEnv, original->getLocation(),
+        original->isBare(), IsNotTransparent, original->isSerialized(),
+        original->isDynamicallyReplaceable());
+    differential->setOwnershipEliminated();
+    differential->setDebugScope(
+        new (module) SILDebugScope(original->getLocation(), differential));
+
+    return differential;
+  }
+
+  /// Run JVP generation. Returns true on error.
+  bool run() {
+    LLVM_DEBUG(getADDebugStream()
+               << "Cloning original @" << original->getName()
+               << " to jvp @" << jvp->getName() << '\n');
+    // Create JVP entry and arguments.
+    auto *entry = jvp->createBasicBlock();
+    createEntryArguments(jvp);
+
+    // Create differential blocks and arguments.
+    // TODO: Consider visiting original blocks in pre-order (dominance) order.
+    SmallVector<SILBasicBlock *, 8> preOrderDomOrder;
+    auto &differential = getDifferential();
+    auto *origEntry = original->getEntryBlock();
+    for (auto &origBB : *original) {
+      auto *diffBB = differential.createBasicBlock();
+      diffBBMap.insert({&origBB, diffBB});
+      auto diffStructLoweredType =
+      remapType(differentialInfo.getLinearMapStructLoweredType(&origBB));
+      // If the BB is the original exit, then the differential block that we just
+      // created must be the differential function's entry. Create differential
+      // entry arguments and continue.
+      if (&origBB == origEntry) {
+        assert(diffBB->isEntry());
+        createEntryArguments(&differential);
+        auto *lastArg = diffBB->getArguments().back();
+        assert(lastArg->getType() == diffStructLoweredType);
+        differentialStructArguments[&origBB] = lastArg;
+      }
+    }
+    assert(diffBBMap.size() == 1 &&
```

Comment:
Why not assert that the function has exactly one block above, and get rid of the for-loop?

Reply:
We will add control flow support and need this for-loop, so I thought I'd keep the for-loop for when this is added in a future patch. You're right in that I don't need it for the current PR.